File: typescript/exercises.ts
import { open } from "node:fs/promises"

export function change(amount: bigint): Map<bigint, bigint> {
  if (amount < 0) {
    throw new RangeError("Amount cannot be negative")
  }
  let counts: Map<bigint, bigint> = new Map()
  let remaining = amount
  for (const denomination of [25n, 10n, 5n, 1n]) {
    counts.set(denomination, remaining / denomination)
    remaining %= denomination
  }
  return counts
}

export function firstThenApply<InputType, OutputType>(
  inputArray: InputType[], 
  condition: (element: InputType) => boolean, 
  transformFunction?: (element: InputType) => OutputType
): OutputType | undefined {
  const matchingElement = inputArray.find(condition); 

  if (matchingElement && transformFunction) {
    return transformFunction(matchingElement);
  }

  return undefined;
}

export function* powersGenerator(base: bigint): Generator<bigint> {
  let power: bigint = 1n; 

  while (true) {
    yield power; 
    power *= base; 
    if (power > 2n ** 100n) break; 
  }
}

export async function meaningfulLineCount(filename: string): Promise<number> {
  let meaningfulLineCount = 0;

  try {
    const file = await open(filename, "r");
    for await (const line of file.readLines()) {
      const trimmedLine = line.trim();
      if (trimmedLine && !trimmedLine.startsWith("#")) {
        meaningfulLineCount++;
      }
    }
  } catch (error) {
    if (error instanceof Error) {
      throw new Error(`Error reading file: ${error.message}`);
    } else {
      throw new Error("Error reading file: Unknown error occurred");
    }
  }

  return meaningfulLineCount;
}

export type Shape = 
    | { kind: "Sphere"; radius: number }
    | { kind: "Box"; width: number; length: number; depth: number };

export function volume(shape: Shape): number {
    switch (shape.kind) {
        case "Sphere":
            return (4 / 3) * Math.PI * Math.pow(shape.radius, 3);
        case "Box":
            return shape.width * shape.length * shape.depth;
        default:
            throw new Error("Shape type not supported for volume calculation");
    }
}

export function surfaceArea(shape: Shape): number {
    switch (shape.kind) {
        case "Sphere":
            return 4 * Math.PI * Math.pow(shape.radius, 2);
        case "Box":
            const { width, length, depth } = shape;
            return 2 * (width * length + width * depth + length * depth);
        default:
            throw new Error("Shape type not supported for surface area calculation");
    }
}

export interface BinarySearchTree<T> {
  size(): number;
  insert(data: T): BinarySearchTree<T>;
  contains(data: T): boolean;
  inorder(): Generator<T>;
}

export class Empty<T> implements BinarySearchTree<T> {
  size(): number {
    return 0;
  }

  insert(data: T): BinarySearchTree<T> {
    return new Node(data, this, this);
  }

  contains(): boolean {
    return false; 
  }

  *inorder(): Generator<T> {
  }

  toString(): string {
    return "()";
  }
  
}

export class Node<T> implements BinarySearchTree<T> {
  constructor(
    private data: T,
    private left: BinarySearchTree<T>,
    private right: BinarySearchTree<T>
  ) {}

  size(): number {
    return 1 + this.left.size() + this.right.size(); 
  }

  insert(data: T): BinarySearchTree<T> {
    if (data < this.data) {
      return new Node(this.data, this.left.insert(data), this.right);
    } else if (data > this.data) {
      return new Node(this.data, this.left, this.right.insert(data));
    }
    return this; 
  }

  contains(data: T): boolean {
    if (data < this.data) {
      return this.left.contains(data);
    } else if (data > this.data) {
      return this.right.contains(data);
    }
    return true; 
  }

  *inorder(): Generator<T> {
    yield* this.left.inorder();  
    yield this.data;            
    yield* this.right.inorder(); 
  }

  toString(): string {
    const leftStr = this.left.toString();
    const rightStr = this.right.toString();
    return `(${leftStr}${this.data}${rightStr})`.replace(/\(\)/g, ""); 
  }
  
}

File: ocaml/exercises.ml
exception Negative_Amount

let change amount =
  if amount < 0 then
    raise Negative_Amount
  else
    let denominations = [25; 10; 5; 1] in
    let rec aux remaining denominations =
      match denominations with
      | [] -> []
      | d :: ds -> (remaining / d) :: aux (remaining mod d) ds
    in
    aux amount denominations

let rec first_then_apply input_list condition transform =
  match input_list with
  | [] -> None
  | current_element :: remaining_elements ->
      if condition current_element then
        Some (transform current_element)
      else
        first_then_apply remaining_elements condition transform


let non_empty s = String.length s > 0;;
let lower s = String.lowercase_ascii s;;
let length_greater_than_3 s = String.length s > 3;;
let square n = n * n;;

let rec powers_generator base =
  let rec aux power =
    fun () ->
      if power > (1 lsl 100) then Seq.Nil
      else Seq.Cons (power, aux (power * base))
  in
  aux 1

let meaningful_line_count (file_name : string) : int =
  try
    let in_channel = open_in file_name in
    let rec count_lines count =
      try
        let line = input_line in_channel in
        let trimmed_line = String.trim line in
        if trimmed_line = "" || String.length trimmed_line > 0 && trimmed_line.[0] = '#' then
          count_lines count
        else
          count_lines (count + 1)
      with
      | End_of_file -> close_in in_channel; count
    in
    count_lines 0
  with
  | Sys_error _ -> raise (Sys_error ("File not found: " ^ file_name))

type shape =
  | Sphere of float  
  | Box of float * float * float 

let volume (s : shape) : float =
  match s with
  | Sphere r -> (4.0 /. 3.0) *. 3.141592653589793 *. r ** 3.0  
  | Box (w, l, d) -> w *. l *. d  

let surface_area (s : shape) : float =
  match s with
  | Sphere r -> 4.0 *. 3.141592653589793 *. r ** 2.0 
  | Box (w, l, d) -> 2.0 *. (w *. l +. w *. d +. l *. d) 

type 'a binary_search_tree =
  | Empty
  | Node of 'a * 'a binary_search_tree * 'a binary_search_tree

let rec size = function
  | Empty -> 0
  | Node (_, left, right) -> 1 + size left + size right

let rec insert data tree =
  match tree with
  | Empty -> Node (data, Empty, Empty)
  | Node (x, left, right) ->
      if data < x then Node (x, insert data left, right)
      else if data > x then Node (x, left, insert data right)
      else tree  

let rec contains data tree =
  match tree with
  | Empty -> false
  | Node (x, left, right) ->
      if data < x then contains data left
      else if data > x then contains data right
      else true  

let rec inorder tree =
  match tree with
  | Empty -> []
  | Node (x, left, right) -> inorder left @ [x] @ inorder right